keps/sig-node/20191127-cri-containerstats-enhancements.md

```diff
@@ -0,0 +1,200 @@
+---
+title: Enhance ContainerStats message in CRI-API
+authors:
+  - "@fuweid"
+owning-sig: sig-node
+participating-sigs:
+  - sig-node
+reviewers:
+  - TBD
+approvers:
+  - TBD
+editor: TBD
+creation-date: 2019-11-27
+last-updated: 2019-11-27
+status: provisional
+---
+
+# Enhance ContainerStats message in CRI-API
+
+## Table of Contents
+
+<!-- toc -->
+- [Summary](#summary)
+- [Motivation](#motivation)
+  - [Goals](#goals)
+  - [Non-Goals](#non-goals)
+- [Proposal](#proposal)
+  - [Implementation Details/Notes/Constraints](#implementation-detailsnotesconstraints)
+- [Design Details](#design-details)
+  - [Test Plan](#test-plan)
+  - [Graduation Criteria](#graduation-criteria)
+  - [Upgrade / Downgrade Strategy](#upgrade--downgrade-strategy)
+  - [Version Skew Strategy](#version-skew-strategy)
+- [Implementation History](#implementation-history)
+<!-- /toc -->
+
+## Summary
+
+This enhancement is to extend `CpuUsage` and `MemoryUsage` fields in
+`ContainerStats`. And also add new `ProcessUsage`/`NetworkUsage` field in
+`ContainerStats`. The Container Runtime Engine will cover more metrics about
+managed container and let the metric collectors focus on node-level metrics or
+monitoring metrics.
+
+## Motivation
+
+Since more and more sandbox container runtimes shows up, like `kata-container`/
+`gVisor`/ `Firecracker`, the way to collect metrics from sandbox container is
```

Comment:
It is probably also worth mentioning windows containers here.  Do they support all of the same metrics as linux containers?

Reply:
Thanks! Windows containers should be mentioned. I updated it.

IIUC, there are two kinds of window containers: one is Linux-based and other one is Windows-based. The metrics from Windows-based one is different from linux's metrics. Not sure that it can be merged into Linux-based. 

For now, the proposal is focus on linux container. And I would like to ask for @Random-Liu @yujuhong @jterry75 @PatrickLang to review it and share options on this.

Ref: https://github.com/microsoft/hcsshim/blob/master/cmd/containerd-shim-runhcs-v1/stats/stats.proto